Tools/CISupport/ews-build/steps.py

```diff
@@ -2255,6 +2271,310 @@ def hideStepIf(self, results, step):
+
+        self.setProperty('list_of_prs', list_of_prs)
+        self.setProperty('all_pr_data', all_pr_data)
+        yield self._addToLog('stdio', 'All PRs in safe-merge-queue: {}\n'.format(list_of_prs))
+        yield self._addToLog('stdio', 'Done!\n')
+
+        defer.returnValue(True)
+
+
+class CheckStatusOfPRsInSafeMergeQueue(buildstep.BuildStep, GitHubMixin, AddToLogMixin):
+    name = 'check-status-of-prs-in-safe-merge-queue'
+    flunkOnFailure = False
+    haltOnFailure = False
+    EMBEDDED_CHECKS = ['ios', 'mac', 'ios-sim', 'ios-wk2', 'ios-wk2-wpt', 'api-ios', 'tv', 'tv-sim', 'watch', 'watch-sim']
+    MACOS_CHECKS = ['mac', 'mac-AS-debug', 'api-mac', 'mac-wk1', 'mac-wk2', 'mac-AS-debug-wk2', 'mac-wk2-stress']
+    QUEUES_FOR_SAFE_MERGE_QUEUE = EMBEDDED_CHECKS + MACOS_CHECKS
+
+    def __init__(self, pr_number='', **kwargs):
+        self.name = f'check-status-of-pr-{pr_number}-in-safe-merge-queue'
+        super().__init__(**kwargs)
+
+    @defer.inlineCallbacks
+    def run(self):
+        pr_number = self.getProperty('pr_number', '')
+        repository = self.getProperty('repository', '')
+        steps_to_add = []
+        checked_status = False
+
+        yield self._addToLog('stdio', f'Performing status check on PR {pr_number}.\n')
+        yield self._addToLog('stdio', f'Link to PR: {repository}/pull/{pr_number}\n')
+
+        # uses result from DetermineLabelOwner
+        committer_status = yield self.validateCommitterStatus(pr_number)
+        if not committer_status:
+            self.setProperty('github.number', pr_number)
+            login = self.getProperty('owners')[0]
+            comment = f'{login} does not have committer permissions.'
+            comment += f'\n\nIf you do have committer permmissions, please ensure that your GitHub username is added to contributors.json.'
+            comment += f'\n\nRejecting commit from merge queue.'
+            self.setProperty('comment_text', comment)
+            yield self._addToLog('stdio', f'Leaving comment on PR {pr_number}...')
+            steps_to_add.append(LeaveComment())
+        else:
+            checked_status = yield self.checkPRStatus(pr_number)
+
+        # after all PRs have been filtered, add corresponding labels
+        if len(self.getProperty('list_of_prs')):
+            steps_to_add += [DetermineLabelOwner()]
+        else:
+            steps_to_add += [AddMergeLabelsToPRs()]
+        self.build.addStepsAfterCurrentStep(steps_to_add)
+
+        defer.returnValue(SUCCESS if checked_status else FAILURE)
+
+    def getResultSummary(self):
+        pr_number = self.getProperty('pr_number')
+        if self.results == SUCCESS:
+            return {'step': f"PR {pr_number} marked safe for merge-queue"}
+        elif self.results == FAILURE:
+            return {'step': f"PR {pr_number} unsafe for merge-queue"}
+        return buildstep.BuildStep.getResultSummary(self)
+
+    @defer.inlineCallbacks
+    def validateCommitterStatus(self, pr_number):
```

Comment:
We're duplicating this code from `MergeQueue` and `UnsafeMergeQueue`, right? Is it possible for us to share it instead?